### PR TITLE
[Timeline] Events methods

### PR DESCRIPTION
**Pull request by ichbinkour**

### CODE REVIEW

File: test/Events/index.test.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import {testGetter} from '../utils/http.js'
+import Events from '../../src/Events'
+
+describe("Events#for", () =>  {
```
Comment (johnsudaar) [inline test/Events/index.test.js]: Pagination should be tested.

File: examples/Addons/list_providers.js [hunk 2]
```diff
@@ -2,7 +2,7 @@ let scalingo = require('../../dist/scalingo.js');
 
 // For database id : 54c6909b61646d0001000000
 scalingo.clientFromToken(process.env.SCALINGO_TOKEN).then((client)=>{
-  return client.Addons.listProviders()
+  return client.Addons.listProviders('postgresql')
```
[the remark this answers is not in the document]
Comment (ichbinkour) [reply examples/Addons/list_providers.js]: This was for a test

File: src/Events/index.js [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+      return unpackData(this._client.apiClient().get(`/apps/${appId}/events`))
+  }
+
+  /**
+   * Return all events for a specific user
+   * @see http://developers.scalingo.com/events#list-current-user-events
+   * @return {Promise<Event | APIError>}
+   */
+  forUser() {
+    return unpackData(this._client.apiClient().get('events'))
```
[the remark this answers is not in the document]
Comment (ichbinkour) [reply src/Events/index.js]: Should I just remove it ?

[the remark this answers is not in the document]
Comment (johnsudaar) [reply src/Events/index.js]: Yeah

File: examples/Events/list_events_for_current_user.js [hunk 4]
```diff
@@ -0,0 +1,9 @@
+let scalingo = require('../../dist/scalingo.js');
+
+scalingo.clientFromToken(process.env.SCALINGO_TOKEN).then((client)=>{
+  return client.Events.forUser()
```
Comment (johnsudaar) [inline examples/Events/list_events_for_current_user.js]: This method is deprecated for now, we should remove it.

File: src/Events/index.js [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+ */
+
+/**
+ * @typedef Event Object of the events and meta data
+ * @property {Events[]} events List of events
+ * @property {PaginationMeta} meta Meta information
+ */
+
+/**
+ * @typedef Pagination Pagination information
```
Comment (johnsudaar) [inline src/Events/index.js]: This already exists in the meta.js file

File: src/Events/index.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (johnsudaar) [reply src/Events/index.js]: UP ?

File: src/Events/index.js [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+   * @param {?Number} from The N last hours - min: 1 max: 72
+   * @param {?Pagination} opts Object that contain the index of the page and the number of element per page
+   * @return {Promise<Event | APIError>}
+   */
+  for(appId, from, opts) {
+    let page
+    let perPage
+
+    if (from && !opts) {
+      return unpackData(this._client.apiClient().get(`/apps/${appId}/events?from${from}`))
```
Comment (johnsudaar) [inline src/Events/index.js]: You're missing a `=` sign here

File: src/Events/index.js [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+    } else if (opts) {
+      page = opts['page'] || 1
+      perPage = opts['per_page'] || 20
+      if (from) 
+        return unpackData(this._client.apiClient().get(`/apps/${appId}/events?from=${from}&page=${page}&per_page=${perPage}`))
```
Comment (johnsudaar) [inline src/Events/index.js]: Use a params object for the query params instead of building the string manually.

File: src/Events/index.js [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+ * @typedef Pagination Pagination information
+ * @property {String} current_page Index of the current page
+ * @property {String} next_page Index of the next page
+ * @property {?String} prev_page Index of the previous page
+ * @property {String} total_pages Total number of pages
+ * @property {String} total_count Total number of items on every pages
+ */
+
+/**
+ * @typedef Events Evet information
```
Comment (johnsudaar) [inline src/Events/index.js]: Why Events ? This is a single Event

File: src/Events/index.js (hunk 6 above)
Comment (ichbinkour) [reply src/Events/index.js]: What do you mean here ?

Comment (johnsudaar) [reply src/Events/index.js]: ~`from${from}`~
`from=${from}`

File: src/Events/index.js (hunk 5 above)
Comment (johnsudaar) [reply src/Events/index.js]: Sooooo ? Delete it ?

File: test/Events/index.test.js [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+  testGetter("https://api.scalingo.com/v1/apps/tata/events", null, null, (client) => {
+    return new Events(client).for("tata", {page: 1, from: 2})
```
Comment (johnsudaar) [inline test/Events/index.test.js]: This is not enough to test if the params (page and per_page) are used

File: src/Events/index.js [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+   */
+  constructor(client) {
+    this._client = client;
+  }
+
+  /**
+   * Return all events of an application
+   * @see http://developers.scalingo.com/events#list-the-events-of-an-app
+   * @param {String} appId Id of the current application
+   * @param {?Number} from The N last hours - min: 1 max: 72
```
Comment (johnsudaar) [inline src/Events/index.js]: This args does note exist...

File: src/Events/index.js [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+    return unpackData(this._client.unauthenticatedClient().get('event_categories'), "event_categories")
+  }
+}
+
+/**
+ * @typedef EventCategory Object of event category
+ * @property {String} id Unique id of event type
+ * @property {String} name Camel case name of the type
+ * @property {String} display_name Fancy name of the type
+ * @property {String} position Order of “importance” when displayed
```
Comment (johnsudaar) [inline src/Events/index.js]: Integer

File: src/Events/index.js [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+ */
+
+/**
+ * @typedef EventType Object of event type
+ * @property {String} id Unique id of event type
+ * @property {String} category_id Category id of event type
+ * @property {String} name Camel case name of the type
+ * @property {String} display_name Fancy name of the type
+ * @property {String} description Description these events are produced
```
Comment (johnsudaar) [inline src/Events/index.js]: This makes no sense.

File: src/Events/index.js [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+ * @typedef Events Object of the events and meta data
+ * @property {Event[]} events List of events
+ * @property {PaginationMeta} meta Meta information
+ */
+
+/**
+ * @typedef Event Event information
+ * @property {String} id Id of the event
+ * @property {Date} created_at Date of the event's creation
+ * @property {User} user Informations of the user
```
Comment (johnsudaar) [inline src/Events/index.js]: This is not a full User object.

File: src/Events/index.js [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+
+/**
+ * @typedef Event Event information
+ * @property {String} id Id of the event
+ * @property {Date} created_at Date of the event's creation
+ * @property {User} user Informations of the user 
+ * @property {String} app_id Id of the application
+ * @property {String} app_name Name of the application
+ * @property {String} type Type of the event
+ * @property {TypeData} type_data Information about the type
```
Comment (johnsudaar) [inline src/Events/index.js]: Object dependant on the event type: https://developers.scalingo.com/events#events

File: src/Events/index.js [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+ * @property {Date} created_at Date of the event's creation
+ * @property {User} user Informations of the user 
+ * @property {String} app_id Id of the application
+ * @property {String} app_name Name of the application
+ * @property {String} type Type of the event
+ * @property {TypeData} type_data Information about the type 
+ */
+
+ /**
+  * @typedef TypeData Type data informations
```
Comment (johnsudaar) [inline src/Events/index.js]: Remove this.

File: src/Events/index.js [hunk 16]
```diff
@@ -0,0 +1,84 @@
+import {unpackData} from '../utils.js'
+import { APIError } from "../errors.js";
+
+/**
+ * Events API Client
+ */
+export default class Tokens {
```
Comment (johnsudaar) [inline src/Events/index.js]: class Events.

File: src/Events/index.js [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+ * @typedef EventType Object of event type
+ * @property {String} id Unique id of event type
+ * @property {String} category_id Category id of event type
+ * @property {String} name Camel case name of the type
+ * @property {String} display_name Fancy name of the type
+ * @property {String} description Description these events are produced
+ */
+
+/**
+ * @typedef Events Object of the events and meta data
```
Comment (johnsudaar) [inline src/Events/index.js]: This might conflict with the class name.

File: src/index.js [hunk 18]
```diff
@@ -94,6 +95,13 @@ class Client {
      * @type {Operation}
      */
     this.Operations = new Operations(this)
+    
+    /**
+     * Events API
+     * @type {Events}
```
Comment (johnsudaar) [inline src/index.js]: This will return the wrong type.

File: test/Events/index.test.js (hunk 9 above)
Comment (johnsudaar) [reply test/Events/index.test.js]: Pagination are still not tested.

File: test/utils/http.js [hunk 19]
```diff
@@ -7,6 +7,35 @@ import {APIError} from '../../src/errors'
 import {Operation} from "../../src/Operations/utils";
 import sinon from "sinon";
 
+export function testEventsGetter(url, opts, build) {
```
Comment (johnsudaar) [inline test/utils/http.js]: This method should not be in `http/utils.js` please move it to `Events/utils.js`.

**Pull request merged**